0000-benchmark-suite.md

```diff
@@ -0,0 +1,782 @@
+DaCapo<sup>[1](#ref-1)</sup> and Renaissance<sup>[2](#ref-2)</sup> benchmark
+suites. It has also been used to show that the SPEC CPU benchmarks includes
+duplicate workloads whose removal would make SPEC CPU smaller and take less time
+to run.<sup>[8](#ref-8)</sup>
+
+The full set of candidates, the metrics instrumentation code, and the PCA
+scripts should all be saved in the benchmarks repository. This way we can
+periodically add a new batch of candidates, recharacterize our corpus, and
+choose a new subset to act as our benchmark suite. We may want to do this, for
+example, once interface types are supported in toolchains and Wasmtime, so we
+can keep an eye on our interface types performance. The full set of candidates
+could also be useful for training PGO builds of Wasmtime and as a corpus from
+which we harvest left-hand sides for offline superoptimization.
+
+### Candidate Program Requirements
+
+* Candidates should be real, widely used programs, or at least extracted kernels
+  of such programs. These programs are ideally taken from domains where Wasmtime
+  and Cranelift are currently used, or domains where they are intended to be a
+  good fit (e.g. serverless compute, game plugins, client Web applications,
+  server Web applications, audio plugins, etc.).
+
+* A candidate program must be deterministic (modulo Wasm nondeterminism like
+  `memory.grow` failure).
+
+* A candidate program must have two associated input workloads: one small and
+  one large. The small workload may be used by developers locally to get quick,
+  ballpark numbers for whether further investment in an optimization is worth
+  it, without waiting for the full, thorough benchmark suite to complete.
+
+* Each workload must have an expected result, so that we can validate executions
+  and avoid accepting "fast" but incorrect results.
+
+* Compiling and instantiating the candidate program and then executing its
+  workload should take *roughly* one to six seconds total.
+
+  > Napkin math: We want the full benchmark to run in a reasonable amount of
+  > time, say twenty to thirty minutes, and we want somewhere around ten to
+  > twenty programs altogether in the benchmark suite to balance diversity,
+  > simplicity, and time spent in execution versus compilation and
+  > instantiation. Additionally, for good statistical analyses, we need *at
+  > least* 30 samples (ideally more like 100) from each benchmark program. That
+  > leaves an average of about one to six seconds for each benchmark program to
+  > compile, instantiate, and execute the workload.
+
+* Inputs should be given through I/O and results reported through I/O. This
+  ensures that the compiler cannot optimize the benchmark program away.
+
+* Candidate programs should only import WASI functions. They should not depend
+  on any other non-standard imports, hooks, or runtime environment.
+
+* Candidate programs must be open source. This makes distributing the benchmark
```

Comment:
+1 for this requirement, I've had bad experiences with closed-source (or not clearly open-source) benchmarks in the past. It would gain being slightly more precise: `open source under a license that allows redistributing, modifying and redistributing modified versions`. This phrasing would allow the most permissive we can get; maybe redistributing is sufficient.

Reply:
Done, thanks.